### PR TITLE
Toggle command with adding roles to ping specific people when a rep is added

### PR DESCRIPTION
**Pull request by night0721**
Added two fields in .env.example so it is customizable which channel to have pings

### CODE REVIEW

**Comment by Zcamt** [comment]
Please change `PING_ID` to `PING_ROLE_ID` so it follows the same format as the other role IDs

**Comment by night0721** [comment]
done

**Comment by Zcamt** [comment]
I would've personally moved this
https://github.com/night0721/discord-bot/blob/9e9244495850caaea5390316a04b22b6f6f611e4/src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt#L126

Alongside where the current message to the support-chat is sent here https://github.com/learnspigot/discord-bot/blob/4a98e5f041772f1653f3d50da35b3e745f15c165/src/main/kotlin/com/learnspigot/bot/entity/UserProfile.kt#L30

@Nuckerr might have a different opinion though, I don't know and I'll let him decide on what he thinks is best. 

Edit:
This change would actually be needed if we also wanted people to be pinged when they receive a rep not specially through a thread, but instead the /managerep add command f.x.

Edit 2: This would remove the need for the `PING_CHANNEL_ID` environmental variable aswell, as we'd just be able to use the support-chat one

**Comment by Nuckerr** [comment]
@night0721 what we need is a mention to support members (literally mentioning the member not a role) in the support channel and deleting the mention afterwards. It would go before/after the message notifying support.

File: .env.example
Comment (Nuckerr) [inline .env.example]: Remove these, use support channel & mention the user

File: src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt]: What is the color? The color used amongst ALL other embeds is added by default in this function and can be accessed by the static const in LearnSpigotBot called `EMBED_COLOR`

File: src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt]: Also this color

File: src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt]: Use mentions of user instead of roles. I would add a boolean to their profile data called `mentionRep` or something and toggle that instead

File: src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt]: Use toggle in profile data

File: src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt
Comment (Nuckerr) [reply src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt]: mb didnt read the code properly, read the latter half of this review

**Comment by night0721** [comment]
@Nuckerr i a mentioning user, or not?

And “It would go before/after the message notifying support.” Like deleting the message after some time?

**Review by Nuckerr — CHANGES_REQUESTED** [review]

**Comment by Nuckerr** [comment]
> @Nuckerr i a mentioning user, or not?
> 
> And “It would go before/after the message notifying support.” Like deleting the message after some time?

You are pinging the user. I misread the code, my bad. Instead of using roles, I would just set a boolean in their user profile.

So you send the message in the function mentioned by @Zcamt 
https://github.com/learnspigot/discord-bot/blob/4a98e5f041772f1653f3d50da35b3e745f15c165/src/main/kotlin/com/learnspigot/bot/entity/UserProfile.kt#L30

and then send the ping message there and and in the .queue consumer, delete it

**Comment by night0721** [comment]
So u mean is using database yeah, since stephen mentioned that using role is fine so I just use roles

**Comment by Nuckerr** [comment]
Ah I see, no worries. I would use the database over roles, less clutter and no issues with discord rate limits

**Comment by night0721** [comment]
I would add the delete timeout and try to do with database when I got back home, I would tell u in discord when I done it

**Comment by Nuckerr** [comment]
Delete messages like this
https://github.com/learnspigot/discord-bot/blob/10e84d5ef22b160932f2533f0a745fdc08d32763/src/main/kotlin/com/learnspigot/bot/command/ThreadCloseCommand.kt#L36

**Comment by night0721** [comment]
done, you may check now

**Comment by night0721** [comment]
done, you may check now

**Comment by night0721** [comment]
done, you may check now

File: .env.example
Comment (Nuckerr) [inline .env.example]: I assume this can be removed now

File: src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt]: I thought we weren't using roles?

**Review by Nuckerr — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt
Comment (Zcamt) [reply src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt]: > I thought we weren't using roles?

I think he's still using roles as that's what Stephen outlined in #47

File: .env.example
Comment (night0721) [reply .env.example]: I didn't add this?

File: src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt
Comment (night0721) [reply src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt]: @Zcamt ^

Comment (Nuckerr) [reply src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt]: @sttephen thoughts? I think it would be a lot cleaner to just keep this in profile data

Comment (Zcamt) [reply src/main/kotlin/com/learnspigot/bot/manager/ForumManager.kt]: > @sttephen thoughts? I think it would be a lot cleaner to just keep this in profile data

Think Stephen mainly suggested using a role to make it a more simple issue to resolve for someone
But I agree with it being cleaner to have it in the profile data

File: .env.example
Comment (night0721) [reply .env.example]: I mean GUILD

**Comment by night0721** [comment]
@Zcamt so will this still be pulled or not

**Comment by Zcamt** [comment]
> @Zcamt so will this still be pulled or not

Currently we'll wait for Stephens response on if he thinks it should go in the profile data or not.
When that's done we can see where we go, if we end up agreeing that a role is fine for now, this will be merged when all, if any, requested changes are done.

And if we decide that we want it in profile data instead, and you're up for it, you can change it and then we can merge

**Comment by Nuckerr** [comment]
> @Zcamt so will this still be pulled or not

Once we sort out all the parts of the code/functionality

**Comment by sttephen** [comment]
@Nuckerr and co, happy for it to be in profile data. I only suggested Roles as it would have saved whoever was going to make the PR (except you) hassle in learning the mongo and profiles setup. But if whoever is making it knows what they're doing, then no prob.

**Comment by Nuckerr** [comment]
@night0721 if you want I can change this out to user profile. I think that it would be a lot cleaner.

**Comment by Zcamt** [comment]
> @Nuckerr and co, happy for it to be in profile data. I only suggested Roles as it would have saved whoever was going to make the PR (except you) hassle in learning the mongo and profiles setup. But if whoever is making it knows what they're doing, then no prob.

I wouldn't personally mind trialing the system with roles at first, and if it proves to be something people use a lot we can always move it over to mongo later 

Also because I don't know how much experience Night has, assuming not much, this would allow for this to be merged and act as a trial for the idea.

But I don't mind either solution, so whatever works for you, works for me

**Comment by sttephen** [comment]
Alright let's run with profiles especially as Nucker has offered to do it, makes me a lot more confident this won't go up in flames!

**Comment by night0721** [comment]
```java
val profile: UserProfile = datastore.findUserProfile(it.user.id)
                val ping: Boolean = profile.pingBoolean
                profile.changePing()
                datastore.save(profile)
                it.replyEmbed({
                    title = "Success"
                    description = "You will ${if (ping) "be" else "not be"} pinged your reputation ping"
                    color = LearnSpigotBot.EMBED_COLOR
                }).queue()
``` 
note: idk will this work

image

**Comment by night0721** [comment]
done @Zcamt @Nuckerr , feel free to check rn

File: src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt
Comment (Zcamt) [inline src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt]: Minor grammatical thing, I think this should be something like "You will be/not be pinged when receiving reputation"

File: src/main/kotlin/com/learnspigot/bot/entity/UserProfile.kt
Comment (Zcamt) [inline src/main/kotlin/com/learnspigot/bot/entity/UserProfile.kt]: In reality, as this is something that gets deleted instantly and only serves as a way to notify the rep-receiver, it doesn't have to have a message outside of it just being `<@${id}>` so I would actually just remove the rest so it's only the mention in the message.

**Review by Zcamt — CHANGES_REQUESTED** [review]
A few changes needed

Edit: Can you also confirm this works and has been tested? :)

**Comment by night0721** [comment]
tested, working

**Comment by Nuckerr** [comment]
Great! Is that tested with an old profile being migrated with this new feature?

File: src/main/kotlin/com/learnspigot/bot/entity/UserProfile.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/entity/UserProfile.kt]: This is essentially a redundant setter. We can just do `profile.reputationPing = !profile.reputationPing` wherever this code is called.

File: src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt
Comment (Nuckerr) [inline src/main/kotlin/com/learnspigot/bot/command/ToggleCommand.kt]: Just do `profile.reputationPing = !profile.reputationPing`

**Review by Nuckerr — CHANGES_REQUESTED** [review]

**Comment by Nuckerr** [comment]
Last couple things and then we can merge!

**Comment by night0721** [comment]
done

**Review by Nuckerr — APPROVED** [review]
Looks good!

**Comment by Nuckerr** [comment]
Will run build & test then merge

**Comment by Nuckerr** [comment]
image

**Comment by Nuckerr** [comment]
@night0721 just make sure you've resolved zcam'ts review

**Review by Zcamt — APPROVED** [review]
Looks good to me 👌

**Comment by night0721** [comment]
@Nuckerr ok now i think

**Pull request merged**
